### PR TITLE
hotplug_mem_repeat: Add new case to test hotplug/unplug 256 times

### PR DESCRIPTION
**Pull request by huangyum**
ID: 1320071
Signed-off-by: Yumei Huang <yuhuang@redhat.com>

### CODE REVIEW

**Comment by huangyum** [comment]
RHEL guest failed the case due to kernel bug. Windows guest don't support hot-unplug memory.
 (1/2) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.0.x86_64.io-github-autotest-qemu.hotplug_memory_repeat: FAIL: Verify unplug memory failed (185.05 s)
 (2/2) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.hotplug_memory_repeat: PASS (258.13 s)

**Comment by huangyum** [comment]
@MiriamDeng Please help  test on ppc host, thanks.

**Comment by yanan-fu** [comment]
@huangyum  remove all the sub test for different police and backend ?  Could you explain the reason ? They are meaningless from product view ?

**Comment by MiriamDeng** [comment]
@huangyum  will it only test default policy in the future ?

**Comment by MiriamDeng** [comment]
@huangyum ,it seems it only supports default policy and basic ram in the future.It looks like the original version support both of them.

**Comment by huangyum** [comment]
@yanan-fu @MiriamDeng I admit it's more comprehensive if we test all possible policy/backend/options.  They are not meaningless but not that meaningful. This case is to test hotplug memory 256 times as it is the number of maximum slots supported. I guess testing with the default policy/backend/option is sufficient to cover it. Besides, for other policy, we need the host has much more memory, especially for ppc host. Many hosts have to cancel most of the tests. So I decided to remove them. Thanks.

File: qemu/tests/hotplug_mem_repeat.py [hunk 1]
```diff
@@ -1,70 +1,58 @@
...
 import logging
 
 from virttest import error_context
 from virttest.utils_test.qemu import MemoryHotplugTest
 
 
 class MemoryHotplugRepeat(MemoryHotplugTest):
-
-    def start_test(self):
+    def repeat_hotplug(self, vm, target_mems):
```
Comment (yanan-fu) [inline qemu/tests/hotplug_mem_repeat.py]: Usually, blank line before function, also after class.

File: qemu/tests/hotplug_mem_repeat.py [hunk 2]
```diff
@@ -1,70 +1,58 @@
...
-        for memory in memorys_added:
-            self.memory_operate(vm, memory, 'offline')
-        time.sleep(1.5)
-        self.unplug_memory(vm, target_mem)
+        Prepare reqired test params, then start memory
+        hotplug memory 256 times and then unplug all of
+        them. And this is test entry.
+        """
+        times = int(self.params["repeat_times"])
+        target_vms = ""
```
Comment (yanan-fu) [inline qemu/tests/hotplug_mem_repeat.py]: Why do not use `target_mems` ?

File: qemu/tests/hotplug_mem_repeat.py [hunk 3]
```diff
@@ -1,70 +1,58 @@
...
-            current_mem = self.get_guest_total_mem(vm)
-            if current_mem != original_mem:
-                self.test.fail("Guest memory changed about repeat"
-                               " hotpug/unplug memory %d times" % repeat)
-            time.sleep(1.5)
-        vm.verify_alive()
-        vm.reboot()
+        for target_mem in target_mems.split():
+            self.hotplug_memory(vm, target_mem)
+        self.check_memory(vm)
```
Comment (yanan-fu) [inline qemu/tests/hotplug_mem_repeat.py]: Why don't check memory after each hotplug ?  and in this way, i am not very sure if it need wait a few seconds to make it take effect, need test.

File: qemu/tests/hotplug_mem_repeat.py [hunk 4]
```diff
@@ -1,70 +1,58 @@
...
-        :param vm: qemu target VM object
-        :param target_mem: memory name of target VM object
-        :param extra_params: params dict, that you want to update
+        Unplug memory in target_mems
+        :param vm: vm object in this test
+        :param target_mems: target memory to be unplugged
+        """
+        for target_mem in target_mems.split():
+            self.unplug_memory(vm, target_mem)
+        self.check_memory(vm)
```
Comment (yanan-fu) [inline qemu/tests/hotplug_mem_repeat.py]: Same with repeat_hotplug.

File: qemu/tests/cfg/hotplug_mem_repeat.cfg [hunk 5]
```diff
@@ -3,44 +3,18 @@
 #    is support memory hotplug;
 - hotplug_memory_repeat:
     type = hotplug_mem_repeat
-    only Linux
     no Host_RHEL.m6
     no RHEL.5
+    no Windows..i386
+    no WinXP Win2000 Win2003 WinVista
     threshold = 210
```
Comment (yanan-fu) [inline qemu/tests/cfg/hotplug_mem_repeat.cfg]: `threshold = 210`?  It is crazy,  it means only if `abs(guest_mem_size - vm_mem_size) > vm_mem_size * threshold`, case fail. refer:  https://github.com/avocado-framework/avocado-vt/blob/e4c4d73b8ec34c6386fcaa1f986767641ab48176/virttest/utils_test/qemu/__init__.py#L709

**Comment by huangyum** [comment]
@yanan-fu I made some changes to keep the original case as well.   Due to bug1684022, the first case would fail, but the second scalability test passed.  Please help review, thanks.
  (2/2) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.0.x86_64.io-github-autotest-qemu.hotplug_memory_repeat.scalability_test: PASS (1462.39 s)

File: qemu/tests/hotplug_mem_repeat.py [hunk 6]
```diff
@@ -1,70 +1,72 @@
...
-        scalability_test = self.params.get("scalability_test") == "yes"
-        repeats = scalability_test and max_slots or self.params["repeats"]
         original_mem = self.get_guest_total_mem(vm)
-        for repeat in range(int(repeats)):
-            extra_params = (scalability_test and
-                            [{'slot_dimm': repeat}] or [None])[0]
-            error_context.context("Hotplug/unplug loop '%d'" % repeat,
-                                  logging.info)
-            self.turn(vm, target_mem, extra_params)
+        if self.params.get("scalability_test") == "yes":
```
Comment (yanan-fu) [inline qemu/tests/hotplug_mem_repeat.py]: We can set a name for `variants`,  then this parameter can be removed.

File: qemu/tests/hotplug_mem_repeat.py [hunk 7]
```diff
@@ -1,70 +1,72 @@
...
 
     def start_test(self):
         """
         Prepare reqired test params, then start memory
-        hotplug/unplug tests in turn. And this is test entry.
+        hotplug memory 256 times and then unplug all of
+        them. And this is test entry.
         """
-        target_mem = self.params["target_mem"]
+        times = int(self.params["repeat_times"])
```
Comment (yanan-fu) [inline qemu/tests/hotplug_mem_repeat.py]: How about `repeats`, it it up to you.

File: qemu/tests/cfg/hotplug_mem_repeat.cfg [hunk 8]
```diff
@@ -3,44 +3,22 @@
...
+    maxmem_mem = 40G
+    ppc64le,ppc64:
+        size_mem = 256M
+        maxmem_mem = 70G
+    backend_mem = memory-backend-ram
+    repeat_times = 256
     variants:
         - @repeat_256:
-            repeats = 256
+            only Linux
```
[the remark this answers is not in the document]
Comment (huangyum) [reply qemu/tests/cfg/hotplug_mem_repeat.cfg]: @yanan-fu Windows guest don't support hot-unplug,  if exclude unplug test, this test is exactly the  same as scalability_test .

File: qemu/tests/hotplug_mem_repeat.py (hunk 7 above)
Comment (huangyum) [reply qemu/tests/hotplug_mem_repeat.py]: I guess both look fine.

File: qemu/tests/hotplug_mem_repeat.py (hunk 3 above)
Comment (huangyum) [reply qemu/tests/hotplug_mem_repeat.py]: We could check memory every time we do hotplug, but it's not that meaningful.  The size for each memdev is only 128M(for x86),  which is a slight change and hardly noticeable due to the threshold.

File: qemu/tests/hotplug_mem_repeat.py (hunk 4 above)
Comment (huangyum) [reply qemu/tests/hotplug_mem_repeat.py]: Same reason as above.

File: qemu/tests/hotplug_mem_repeat.py (hunk 1 above)
Comment (huangyum) [reply qemu/tests/hotplug_mem_repeat.py]: Thanks, updated.

File: qemu/tests/hotplug_mem_repeat.py [hunk 9]
```diff
@@ -7,64 +6,68 @@
...
+        if self.params["test_type"] == "scalability_test":
+            error_context.context("Repeat hotplug memory for %s times"
+                                  % times, logging.info)
+            self.repeat_hotplug(vm, target_mems)
+            if self.params.get('os_type') == 'linux':
+                error_context.context("Repeat unplug memory for %s times"
+                                      % times, logging.info)
+                self.repeat_unplug(vm, target_mems)
+        else:
+            for target_mem in target_mems.split():
```
Comment (yanan-fu) [inline qemu/tests/hotplug_mem_repeat.py]: We can convert `target_mems` to a list after it is generated, then no need to convert it for again in `repeat_hotplug`, `repeat_unplug` .

File: qemu/tests/hotplug_mem_repeat.py (hunk 3 above)
Comment (yanan-fu) [reply qemu/tests/hotplug_mem_repeat.py]: @huangyum  It is ok,  as i just realized `scalability_test` and `repeat_256` check memory at different timing (check after all mem be pluged  and check after every one be pluged), both of these scenario can be covered.
My concern is,  for `repeat_256`,  it do check memory after plug/unplug one memdev (128M), is it ok as you mentioned it is hardly noticeable due to the threshold ?   do me need to adjust threshold ?

Comment (yanan-fu) [reply qemu/tests/hotplug_mem_repeat.py]: @huangyum  correct the comment above,  no `check_memory` for `repeat_256`.  it use  `hotplug_memory`, `unplug_memory` directly. ignore it.

File: qemu/tests/cfg/hotplug_mem_repeat.cfg (hunk 8 above)
[the remark this answers is not in the document]
Comment (yanan-fu) [reply qemu/tests/cfg/hotplug_mem_repeat.cfg]: @huangyum  ahah, yes,  they are same.

**Comment by yanan-fu** [comment]
`scalability_test` is ok for me, as we discussed, it hotplug 256 mem_devs in total, will find free `slot` automatically from 0 to 255.    
For `repeat_256`,  I prefer hotplug mem_devs to different slot instead of the fixed slot 0. As i think it can cover more scenarios.  It is up to you if it is meaningless from product view.

File: qemu/tests/hotplug_mem_repeat.py (hunk 9 above)
Comment (huangyum) [reply qemu/tests/hotplug_mem_repeat.py]: Good suggestion! Have updated to initialize it as a list.

**Comment by huangyum** [comment]
> `scalability_test` is ok for me, as we discussed, it hotplug 256 mem_devs in total, will find free `slot` automatically from 0 to 255.
> For `repeat_256`, I prefer hotplug mem_devs to different slot instead of the fixed slot 0. As i think it can cover more scenarios. It is up to you if it is meaningless from product view.

The test for slots should have been covered by `scalability_test`,  `repeat_256` mainly focus on the stress test, it should not care which slot to plugged in.

**Comment by huangyum** [comment]
Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.0.x86_64.io-github-autotest-qemu.hotplug_memory_repeat.scalability_test: PASS (679.67 s)

**Comment by huangyum** [comment]
@MiriamDeng  Would you please have a try on ppc host? Case would fail due to the unplug bug on ppc, but  it can somehow verify the error message. Thanks!

**Comment by MiriamDeng** [comment]
For the jobs,they were like what huangyum had mentioned it will fail because of known bug.
After adding parameter to cmdline,Still got failures.Consulted with huangyum,code will be updated soon.Thanks.
(1/2) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.0.ppc64le.io-github-autotest-qemu.hotplug_memory_repeat.repeat_256: FAIL: Verify unplug memory failed (100.78 s)
 (2/2) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.0.ppc64le.io-github-autotest-qemu.hotplug_memory_repeat.scalability_test: FAIL: Assigned '4096 MB' memory to 'avocado-vt-vm1'but, '3547 MB' memory detect by OS (675.07 s)

**Comment by huangyum** [comment]
@yanan-fu Would you please help review again, I just added threshold for ppc and update kernel line option for guest. Thanks!
@MiriamDeng Please try again, the issue should have been fixed. Thanks.

**Comment by yanan-fu** [comment]
@huangyum  without the arg, linux guest support hot-unplug too, but failed sometimes as a product bug(from my test result before). 
So, why add this arg ?  And i see in the comment from MiriamDenf "After adding parameter to cmdline,Still got failures."
Could you explain with more details,  and also add the bug id if it is ok to add it here. Thanks !

**Comment by MiriamDeng** [comment]
@yanan-fu,the arg just can improve the success ratio while doing hot-unplug memory.About this failure"After adding parameter to cmdline,Still got failures.",it was a problem of threshold while doing memory check,which I discussed it with @huangyum yesterday.
So far,the latest results from ppc part.
 (1/2) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.0.ppc64le.io-github-autotest-qemu.hotplug_memory_repeat.repeat_256: PASS (613.92 s)
 (2/2) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.0.ppc64le.io-github-autotest-qemu.hotplug_memory_repeat.scalability_test: PASS (679.55 s)

**Comment by MiriamDeng** [comment]
@yanan-fu,the arg just can improve the success ratio while doing hot-unplug memory.About this failure"After adding parameter to cmdline,Still got failures.",it was a problem of threshold while doing memory check,which I've discussed with @huangyum yesterday.
So far,the latest results from ppc part.
(1/2) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.0.ppc64le.io-github-autotest-qemu.hotplug_memory_repeat.repeat_256: PASS (613.92 s)
(2/2) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.0.ppc64le.io-github-autotest-qemu.hotplug_memory_repeat.scalability_test: PASS (679.55 s)

**Comment by yanan-fu** [comment]
Then the issues are:
1. hot-unplug with linux guest failed as product bz( 1245892 ? )
2. add kernel arg `movable_node` in guest kernel line, can reduce the probability of failure in 1 ?   or "Memory hot-unplug only works after adding 'movable_node' to guest kernel line." (i see it with patch https://github.com/autotest/tp-qemu/pull/1692)  ?
3. threshold issue that already be fixed in the latest update ?
@huangyum  @MiriamDeng

**Comment by MiriamDeng** [comment]
@ yanan-fu,3 is ppc specific problem and now it should be resolved.Thanks.

**Comment by huangyum** [comment]
> @yanan-fu I made some changes to keep the original case as well. Due to bug1684022, the first case would fail, but the second scalability test passed. Please help review, thanks.
> (2/2) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.0.x86_64.io-github-autotest-qemu.hotplug_memory_repeat.scalability_test: PASS (1462.39 s)

@yanan-fu They are different bugs.  Pls refer to the bug mentioned in above comment. Thanks.

**Comment by yanan-fu** [comment]
Ack again.

**Pull request merged**